src/util/side_metadata/helpers.rs

```diff
@@ -0,0 +1,210 @@
+use crate::util::{constants, memory, Address};
+use memory::dzmmap;
+use std::io::{Error, Result};
+
+use super::global::METADATA_SINGLETON;
+use super::SideMetadataID;
+
+#[cfg(target_pointer_width = "32")]
+pub(super) const METADATA_BASE_ADDRESS: Address = unsafe { Address::from_usize(0) };
+#[cfg(target_pointer_width = "64")]
+pub(super) const METADATA_BASE_ADDRESS: Address =
+    unsafe { Address::from_usize(0x0000_0600_0000_0000) };
+
+#[cfg(target_pointer_width = "32")]
+pub(super) const MAX_HEAP_SIZE_LOG: usize = 32;
+// FIXME: This must be updated if the heap layout changes
+#[cfg(target_pointer_width = "64")]
+pub(super) const MAX_HEAP_SIZE_LOG: usize = 48;
+
+pub(super) const MAX_METADATA_BITS: usize = constants::BITS_IN_WORD;
+// const SPACE_PER_META_BIT: usize = 2 << (MAX_HEAP_SIZE_LOG - constants::LOG_BITS_IN_WORD);
+pub(super) const META_SPACE_PAGE_SIZE: usize = constants::BYTES_IN_PAGE;
+pub(super) const META_SPACE_PAGE_SIZE_LOG: usize = constants::LOG_BYTES_IN_PAGE as usize;
+
+#[inline(always)]
+pub(super) fn address_to_meta_address(addr: Address, metadata_id: SideMetadataID) -> Address {
+    let bits_num_log = METADATA_SINGLETON.meta_bits_num_log_vec[metadata_id.as_usize()] as i32;
+    // right shifts for `align` times, then
+    // if bits_num_log < 3, right shift a few more times to cover multi objects per metadata byte
+    // if bits_num_log = 3, metadata byte per object is 1
+    // for > 3, left shift, because more than 1 byte per object is required
+    let rshift = (constants::LOG_BITS_IN_BYTE as i32) - bits_num_log;
+    let offset = if rshift >= 0 {
+        addr.as_usize()
+            .checked_shr(METADATA_SINGLETON.align[metadata_id.as_usize()] as u32)
+            .unwrap()
+            .checked_shr(rshift as u32)
+            .unwrap()
```

Comment:
Probably you can simply use >> instead of checked_str. Because you are doing unwrap() to the result anyway.

Using debug_assert and >> to replace checked_str can also reduce runtime checks for release builds.

Reply:
Right, the reason why I ended up using the checked version of shr/shl was that I was trying not to use the if statement `if rshift >= 0` and was getting errors from `>>`.
In the end I couldn't do what I wanted and also forgot to revert the changes from `>>` to `checked_shr` 😄